src/bootstrap/compile.rs

```diff
@@ -1108,6 +1108,11 @@ impl Step for Assemble {
             let src_exe = exe("lld", target_compiler.host);
             let dst_exe = exe("rust-lld", target_compiler.host);
             builder.copy(&lld_install.join("bin").join(&src_exe), &libdir_bin.join(&dst_exe));
+            // for gcc-lld linker flavor
+            builder.copy(
+                &lld_install.join("bin").join(&src_exe),
+                &libdir_bin.join(exe("ld", target_compiler.host)),
```

Comment:
The comment above says:
```rust
// We prepend this bin directory to the user PATH when linking Rust binaries. To
// avoid shadowing the system LLD we rename the LLD we provide to `rust-lld`.
```
Namely, the `lib/rustlib/<host>/bin` directory is added to `PATH` in `fn get_linker`.

This means `ld` (or `ld.exe`) will interfere with gcc's linker search even if `-Zgcc-ld=lld` is not passed.

GCC prefers hardcoded paths like `/mingw64/x86_64-w64-mingw32/bin` (on windows-gnu) to searching `PATH` so the right ld was chosen in my case, but this newly added `ld` can still interfere with search in other cases.

Reply:
I think this can be addressed by placing `ld`/`ld.exe` to a new subdirectory in `lib/rustlib/<host>/bin` (similarly to `lib/rustlib/<host>/bin/self-contained`), which is never added to `PATH` and is only used in `-Bprefix` when `-Z gcc-ld=lld` is enabled.